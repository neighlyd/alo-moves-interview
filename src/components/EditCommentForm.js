import React, { useState } from 'react';
import Modal from 'react-modal'

const CommentForm = (props) => {

  const [ comment, setComment ] = useState(props.comment);

  const origComment = props.comment;

  const handleSubmit = e => {
    e.preventDefault();
    if (comment !== '' && comment !== origComment) {
      props.handleSubmit({text: comment});
    }
  };

  return (
    <Modal
      isOpen={!!props.modalOpen}
      className="modal"
      closeTimeoutMS={200}
      contentLabel={props.formTitle}
      onRequestClose={props.handleModalClose}
    >
      <div>
        <h3>{props.formTitle}</h3>
        <form onSubmit={handleSubmit} className='edit-comment'>
          <label
            className='edit-comment__label'
          >
            Comment
          </label>
          <textarea 
            rows='4'
            className='edit-comment__input text-input'
            autoFocus
            value={comment}
            onChange={e => setComment(e.target.value)}
          />
          { (comment !== '' && comment !== props.comment) ? (
            <button 
              className='edit-comment__btn'
            >
              {props.buttonLabel}
            </button> 
          ) : (
            <button 
              disabled
              className='edit-comment__btn'
            >
              {props.buttonLabel}
            </button>
          )}
        </form>
      </div>
    </Modal>
  )
}

export default CommentForm;