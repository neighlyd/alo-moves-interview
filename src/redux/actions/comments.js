import firebase from '../../firebase/firebase';
const db = firebase.collection('comments')

// Use const variables for action names, so we have a single place to edit them in the future.
export const ADD_COMMENT = 'ADD_COMMENT';
export const DELETE_COMMENT = 'DELETE_COMMENT';
export const EDIT_COMMENT = 'EDIT_COMMENT';
export const SET_COMMENTS = 'SET_COMMENTS';

// These are the pure functions that will be dispatched to the Redux store itself.
export const addComment = (comment) => ({
  type: ADD_COMMENT,
  comment
});

export const deleteComment = (id) => ({
  type: DELETE_COMMENT,
  id
});

export const editComment = (id, updates) => ({
  type: EDIT_COMMENT,
  id,
  updates
});

export const setComments = (comments) => ({
  type: SET_COMMENTS,
  comments
})

export const startAddComment = (comment = '') => {
  return async dispatch => {  
    try {      
      const data = {
        text: comment,
        updatedAt: new Date()
      };
      // Firestore's .add() method returns a documentReference. We need to populate this reference with the newly created object using the .get() method.
      // Once we have a populated object, we can then dispatch it to our Redux store.
      const docRef = await db.add(data);
      const doc = await docRef.get();
      
      if (doc) {
        dispatch(addComment({
          id: doc.id,
          text: doc.data().text,
          updatedAt: doc.data().updatedAt
        }));
      }
    } catch (err){
      console.log(err);
    };
  };
};

export const startDeleteComment = id => {
  return async dispatch => {  
    try{
      
      // First try to retrieve the Document from the Firestore DB.
      // If it still exists there, delete it remotely and locally.
      // If it has already been removed from remote, then only remove from local (e.g., another user deleted the comment and the page hasn't refreshed yet)
      // The need for this check could be partially alleviated by using sockets or subscribing to the DB store itself.
      const docRef = await db.doc(id).get();
      
      if (docRef.exists) {
        await db.doc(id).delete();
      } 
      dispatch(deleteComment(id));
    } catch (err) {
      console.log(err);
    };
  };
};

export const startEditComment = (id, comment) => {
  return async dispatch => {
    try {
      // As above, we error check to see if the document has been deleted by another user.
      const docRef = await db.doc(id).get();
      if (!docRef.exists) {
        // If another user deleted the comment before our user refreshes the page and attempts to edit, then their local version should be removed as well.
        // Otherwise, our local store will be out of sync.
        dispatch(deleteComment(id));
      } else {
        const updates = {
          ...comment,
          updatedAt: new Date()
        }
        await db.doc(id).set(updates)

        // Firestore does not return a document upon set(), so we need to retrieve the updated document from the remote DB.
        // We need to do this because they auto-format timestamp fields in their own format, which is incompatible with javascript's native Date() type.
        const doc = await db.doc(id).get();
        
        if (doc.exists) {
          const updates = {
            text: doc.data().text,
            updatedAt: doc.data().updatedAt
          };
        
          dispatch(editComment(doc.id, updates));
        } 
      }
    } catch (err) {
      console.log(err);
    };
  };
};

export const startSetComments = () => {
  return async dispatch => {
    try {
      
      // Start by fetching all comments from Firestore.
      // we then need to parse the data into an array (Firestore doesn't store info in arrays - at least, it's a pain to do so)
      // Finally, we dispatch() SET_EXPENSES with the new data array.
      let comments = [];
      
      const docRef = await db.get()
      
      docRef.forEach(comment => {
        comments.push({
          id: comment.id,
          text: comment.data().text,
          updatedAt: comment.data().updatedAt
        });
      });
      
      dispatch(setComments(comments));
    
    } catch (err) {
      console.log(err);
    };
  };
};