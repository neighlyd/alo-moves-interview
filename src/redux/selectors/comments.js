export default (comments) => {
  return comments.sort((a,b) => {
    return a.updatedAt < b.updatedAt ? -1 : 1;
  });
};