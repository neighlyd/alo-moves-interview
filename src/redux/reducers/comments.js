import { ADD_COMMENT, DELETE_COMMENT, EDIT_COMMENT, SET_COMMENTS } from '../actions/comments';

// Comments will be an array of objects. As such:
// let comments = [{id: 1, text: 'comment 1'}, {id: 2, text: 'comment 2'}]

const commentsReducerDefaultState = [];

export default (state = commentsReducerDefaultState, action) => {
  switch(action.type){
    case ADD_COMMENT:
      return [
        ...state,
        action.comment
      ];
    case DELETE_COMMENT:
      return state.filter( comment => comment.id !== action.id);
    case EDIT_COMMENT:
      return state.map(comment => {
        if (comment.id === action.id){
          return {
            ...comment,
            ...action.updates
          }
        } else {
          return comment;
        }
      });
    case SET_COMMENTS:
      return action.comments
    default:
      return state;
  };
};