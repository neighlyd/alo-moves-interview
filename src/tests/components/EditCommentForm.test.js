import React from 'react';
import { shallow } from 'enzyme';

import EditCommentForm from '../../components/EditCommentForm';
import comments from '../fixtures/comments';

let comment, handleSubmit, wrapper;

beforeEach(() => {
  comment = comments[0];
  handleSubmit = jest.fn();
  wrapper = shallow(<EditCommentForm comment={comment.text} handleSubmit={handleSubmit}/>);
});

it('EditCommentForm should render correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

it('EditCommentForm should handleSubmit correctly', () => {
  const newComment = 'New Comment';
  const input = wrapper.find('.edit-comment__input')
  input.simulate('change', {target: {value: newComment}});

  wrapper.find('.edit-comment').simulate('submit', {preventDefault: () => null});
  expect(handleSubmit).toHaveBeenCalledWith({text: newComment})
});