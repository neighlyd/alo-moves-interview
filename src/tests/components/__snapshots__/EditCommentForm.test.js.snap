// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditCommentForm should render correctly 1`] = `ShallowWrapper {}`;
